File: java_app/src/main/java/me/halin/android_architecture/uikit/DataBindingAdapter.java
package me.halin.android_architecture.uikit;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * 抽离模型与UI,所有的UI操作封装到Factory中实现
 */
public abstract class DataBindingAdapter<T> extends RecyclerView.Adapter<DataBindingHolder> {


    public final DataBindingHolder.Factory<T> holderFactory;

    public DataBindingAdapter(DataBindingHolder.Factory<T> holderFactory) {
        this.holderFactory = holderFactory;
    }

    @NonNull
    @Override
    public DataBindingHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return holderFactory.createHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull final DataBindingHolder holder, int position) {
        holder.bindTo(getModelAtPosition(position));
    }


    @Override
    public int getItemViewType(int position) {
        T model = getModelAtPosition(position);
        return this.holderFactory.getViewType(model);
    }

    public abstract T getModelAtPosition(int position);


}

File: app/src/main/java/me/halin/android_architecture/uikit/DataBindingAdapter.kt
package me.halin.android_architecture.uikit

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

public abstract class DataBindingAdapter<T>(val holderFactory: DataBindingHolder.HolderFactory<T>) : RecyclerView.Adapter<DataBindingHolder>() {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DataBindingHolder {
        return holderFactory.createHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: DataBindingHolder, position: Int) {
        holder.bindTo(getModelAtPosition(position))
    }

    override fun getItemViewType(position: Int): Int {
        return holderFactory.getViewType(getModelAtPosition(position))
    }


    public abstract fun getModelAtPosition(position: Int): T
}